File: legions/shell_executor.py
#!/usr/bin/env python3
"""
ROME Shell Executor v2.0
Directly executes bash commands, captures output, and reports progress.
"""
import sys
import os
import time
import subprocess
import json
import threading

# Silently import _log_event from rome_log if available.
try:
    # Attempt to find rome_log in its likely location: ../dictator/rome_log.py
    sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'dictator'))
    from rome_log import _log_event
except ImportError:
    _log_event = None

# Silently import ws_client from ws_client if available.
try:
    # Attempt to find ws_client in its likely location: ../dictator/ws_client.py
    sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'dictator'))
    from ws_client import send_event, flush
except ImportError:
    send_event = None
    flush = None

def shell_executor():
    # Priority: Env > Command Line > Default
    task_id = os.environ.get("ROME_TASK_ID")
    if not task_id:
        if len(sys.argv) > 1:
            task_id = sys.argv[1]
        else:
            task_id = f"shell_{int(time.time())}"

    # Priority: Env > Command Line > Current Time
    global_start_timestamp = None
    if os.environ.get("ROME_START_TIME"):
        try:
            global_start_timestamp = float(os.environ.get("ROME_START_TIME"))
        except: pass
    
    if global_start_timestamp is None:
        if len(sys.argv) > 2:
            try:
                global_start_timestamp = float(sys.argv[2])
            except: pass
    
    if global_start_timestamp is None:
        global_start_timestamp = time.time()
    
    # Priority: Last Argument > Env
    bash_command = None
    if len(sys.argv) >= 2:
        bash_command = sys.argv[-1]
    
    if not bash_command:
        bash_command = os.environ.get("ROME_PROMPT")

    if not bash_command:
        print("Usage: python3 shell_executor.py <task_id> <global_start_timestamp> <bash_command_string>")
        print("OR set ROME_TASK_ID and pass bash_command as last arg.")
        sys.exit(1)

    shell_timeout = float(os.environ.get('SHELL_TIMEOUT', 600))

    # task_dir = <rome-core>/legions/<task_id>/ — create if needed
    base_dir = os.path.dirname(os.path.abspath(__file__))
    task_dir = os.path.join(base_dir, task_id)
    os.makedirs(task_dir, exist_ok=True)

    start_time = time.time()
    full_output = []
    
    # Run: subprocess.Popen(['bash', '-c', bash_command], stdout=PIPE, stderr=STDOUT, text=True)
    process = subprocess.Popen(
        ['bash', '-c', bash_command],
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
        bufsize=1,
        start_new_session=True,
    )

    def read_output(proc, output_list, tid, start_t):
        # Read stdout line by line (BLOCKING — no O_NONBLOCK)
        # Using iter(proc.stdout.readline, '') to ensure blocking reads.
        for line in iter(proc.stdout.readline, ''):
            output_list.append(line)
            last_line = line.strip()
            elapsed = time.time() - start_t
            # Print per-line progress to own stdout: PROGRESS:{task_id}:{elapsed:.1f}s:{last_line[:80]}
            print(f"PROGRESS:{tid}:{elapsed:.1f}s:{last_line[:80]}", flush=True)

            # Send real-time progress via WebSocket if available (throttled)
            if send_event and (len(output_list) % 5 == 0 or elapsed < 2):
                # We don't have a real total, so we use a rolling 'pseudo-progress'
                # that increments with line count, capped at 99%.
                percent = min(1 + (len(output_list) // 5), 99)
                send_event("progress", tid, {"percent": float(percent), "message": last_line[:80]})
        proc.stdout.close()

    reader_thread = threading.Thread(target=read_output, args=(process, full_output, task_id, start_time))
    reader_thread.daemon = True
    reader_thread.start()

    # Wait for process to finish or timeout
    is_timeout = False
    while process.poll() is None:
        if time.time() - start_time > shell_timeout:
            # Kill entire process group to avoid orphans
            import signal
            try:
                os.killpg(os.getpgid(process.pid), signal.SIGKILL)
            except OSError:
                process.kill()
            is_timeout = True
            break
        time.sleep(0.05)

    # Ensure reader thread finishes
    reader_thread.join(timeout=1.0)

    exit_code = process.returncode if not is_timeout else -1
    status = "SUCCESS" if exit_code == 0 and not is_timeout else "FAILED"
    elapsed_s = time.time() - start_time

    # Append timeout message if it happened
    if is_timeout:
        full_output.append(f"\n[TIMEOUT] Process killed after {shell_timeout}s\n")

    # Write full output to task_dir/report_{task_id}.txt
    report_path = os.path.join(task_dir, f"report_{task_id}.txt")
    try:
        with open(report_path, "w") as f:
            f.writelines(full_output)
    except Exception:
        pass

    # Write manifest.json: {rome_v:"2.0", task_id, status, usage:null, runtime:{elapsed_s, exit_code}}
    manifest = {
        "rome_v": "2.0",
        "task_id": task_id,
        "status": status,
        "usage": None,
        "runtime": {
            "elapsed_s": round(elapsed_s, 3),
            "exit_code": exit_code
        }
    }
    manifest_path = os.path.join(task_dir, "manifest.json")
    try:
        with open(manifest_path, "w") as f:
            json.dump(manifest, f, indent=2)
    except Exception:
        pass

    if send_event:
        send_event("complete", task_id, {"status": status, "report_path": report_path, "usage": None})
        if flush:
            flush()

    # Print last 3 lines as summary
    print("\n--- Summary ---")
    summary_lines = full_output[-3:]
    for l in summary_lines:
        print(l.strip())



if __name__ == "__main__":
    shell_executor()

File: src/shell_executor.ts
import { spawn } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import { EventEmitter } from 'events';

const ROME_ROOT = process.env.ROME_ROOT || process.cwd();

export function executeShell(
  taskId: string,
  command: string,
  wsSender?: (ev: any) => Promise<void>,
  cancelEmitter?: EventEmitter
): Promise<{ status: 'SUCCESS' | 'FAILED'; report: string; exit_code: number; elapsed_s: number }> & { kill?: () => void } {
  const taskDir = path.join(ROME_ROOT, 'legions', taskId);
  if (!fs.existsSync(taskDir)) {
    fs.mkdirSync(taskDir, { recursive: true });
  }

  const startTs = Date.now();
  const chunks: Buffer[] = [];
  let interrupted = false;
  let interruptReason = '';
  let lineCount = 0;
  let totalLineCount = 0;
  let lastProgressTs = Date.now();
  let proc!: ReturnType<typeof spawn>;

  const promise = new Promise<{ status: 'SUCCESS' | 'FAILED'; report: string; exit_code: number; elapsed_s: number }>((resolve) => {
    proc = spawn('bash', ['-c', command], {
      detached: true,
      cwd: taskDir,
      env: {
        ...process.env,
        ROME_TASK_ID: taskId,
        ROME_TASK_DIR: taskDir
      }
    });

    const handleCancel = (id: string) => {
      if (id === taskId) {
        interrupted = true;
        interruptReason = 'CANCEL';
        try { process.kill(-proc.pid!, 'SIGKILL'); } catch (e) {}
      }
    };

    const handleTimeout = (id: string) => {
      if (id === taskId) {
        interrupted = true;
        interruptReason = 'TIMEOUT';
        try { process.kill(-proc.pid!, 'SIGKILL'); } catch (e) {}
      }
    };

    if (cancelEmitter) {
      cancelEmitter.on('cancel', handleCancel);
      cancelEmitter.on('timeout', handleTimeout);
    }

    const onData = (data: Buffer) => {
      chunks.push(data);
      const output = data.toString('utf-8');
      const lines = output.split('\n');
      lineCount += lines.length - 1;
      totalLineCount += lines.length - 1;
      const now = Date.now();
      const elapsed = (now - startTs) / 1000;

      if (wsSender && (lineCount >= 5 || elapsed < 2 || now - lastProgressTs >= 2000)) {
        const lastLine = lines[lines.length - 1] || lines[lines.length - 2] || '';
        const message = lastLine.trim().slice(0, 80);
        const percent = Math.min(1 + Math.floor(totalLineCount / 5), 99);

        wsSender({
          type: 'progress',
          task_id: taskId,
          payload: {
            percent,
            message
          }
        }).catch(() => {});
        lineCount = 0;
        lastProgressTs = now;
      }
    };

    proc.stdout?.on('data', onData);
    proc.stderr?.on('data', onData);

    const cleanup = () => {
      if (cancelEmitter) {
        cancelEmitter.removeListener('cancel', handleCancel);
        cancelEmitter.removeListener('timeout', handleTimeout);
      }
    };

    proc.on('close', (code: number) => {
      cleanup();
      if (interrupted) {
          chunks.push(Buffer.from(`\n[${interruptReason}] Process killed\n`));
      }
      const elapsed_s = (Date.now() - startTs) / 1000;
      const status = (code === 0 && !interrupted) ? 'SUCCESS' : 'FAILED';
      const report = Buffer.concat(chunks).toString('utf-8').slice(0, 4000);

      resolve({
        status,
        report,
        exit_code: code ?? -1,
        elapsed_s: Math.round(elapsed_s * 1000) / 1000
      });
    });

    proc.on('error', (err: Error) => {
      cleanup();
      const elapsed_s = (Date.now() - startTs) / 1000;
      chunks.push(Buffer.from(`\n[ERROR] ${err.message}\n`));
      resolve({
        status: 'FAILED',
        report: Buffer.concat(chunks).toString('utf-8').slice(0, 4000),
        exit_code: -1,
        elapsed_s: Math.round(elapsed_s * 1000) / 1000
      });
    });
  });

  (promise as any).kill = () => {
    if (proc && proc.pid) {
      try { process.kill(-proc.pid, 'SIGTERM'); } catch (e) {}
    }
  };

  return promise;
}
